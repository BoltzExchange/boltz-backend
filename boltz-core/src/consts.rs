pub const ECDSA_BYTES_TO_GRIND: usize = 2;

pub const PREIMAGE_DUMMY: [u8; 0] = [];

pub const STUB_SCHNORR_SIGNATURE_LENGTH: usize = 64;
