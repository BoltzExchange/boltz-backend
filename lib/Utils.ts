import os from 'os';
import path from 'path';
import { OutputType, Scripts } from 'boltz-core';

const { p2wshOutput, p2shP2wshOutput, p2shOutput, p2wpkhOutput, p2pkhOutput, p2shP2wpkhOutput } = Scripts;

/**
 * Splits a derivation path into multiple parts
 */
export const splitDerivationPath = (path: string): { master: string, sub: number[] } => {
  const split = path.split('/');
  const master = split.shift()!;

  const sub: number[] = [];

  split.forEach((part) => {
    sub.push(Number(part));
  });

  return {
    master,
    sub,
  };
};

/**
 * Concat an error code and its prefix
 */
export const concatErrorCode = (prefix: number, code: number) => {
  return `${prefix}.${code}`;
};

/**
 * Capitalize the first letter of a string
 */
export const capitalizeFirstLetter = (input: string) => {
  return input.charAt(0).toUpperCase() + input.slice(1);
};

/**
 * Resolve '~' on Linux and Unix-Like systems
 */
export const resolveHome = (filename: string) => {
  if (os.platform() !== 'win32') {
    if (filename.charAt(0) === '~') {
      return path.join(process.env.HOME!, filename.slice(1));
    }
  }

  return filename;
};

/**
 * Get a hex encoded Buffer from a string
 * @returns a hex encoded Buffer
 */
export const getHexBuffer = (input: string) => {
  return Buffer.from(input, 'hex');
};

/**
 * Get a hex encoded string from a Buffer
 *
 * @returns a hex encoded string
 */
export const getHexString = (input: Buffer) => {
  return input.toString('hex');
};

/**
 * Check whether a variable is a non-array object
 */
export const isObject = (val: any): boolean => {
  return (val && typeof val === 'object' && !Array.isArray(val));
};

/**
 * Get the current date in the LocaleString format.
 */
export const getTsString = (): string => (new Date()).toLocaleString('en-US', { hour12: false });

/**
 * Recursively merge properties from different sources into a target object, overriding any
 * existing properties.
 * @param target The destination object to merge into.
 * @param sources The sources objects to copy from.
 */
export const deepMerge = (target: any, ...sources: any[]): object => {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    Object.keys(source).forEach((key) => {
      if (isObject(source[key])) {
        if (!target[key]) Object.assign(target, { [key]: {} });
        deepMerge(target[key], source[key]);
      } else if (source[key] !== undefined) {
        Object.assign(target, { [key]: source[key] });
      }
    });
  }

  return deepMerge(target, ...sources);
};

/**
 * Get all methods from an object whose name doesn't start with an underscore.
 */
export const getPublicMethods = (obj: any): any => {
  const ret = {};
  Object.getOwnPropertyNames(Object.getPrototypeOf(obj)).forEach((name) => {
    const func = obj[name];
    if ((func instanceof Function) && name !== 'constructor' && !name.startsWith('_')) {
      ret[name] = func;
    }
  });
  return ret;
};

export const groupBy = (arr: object[], keyGetter: (item: any) => string | number): any => {
  const ret = {};
  arr.forEach((item) => {
    const key = keyGetter(item);
    const group = ret[key];
    if (!group) {
      ret[key] = [item];
    } else {
      group.push(item);
    }
  });
  return ret;
};

/**
 * Get current time in unix time (milliseconds).
 */
export const ms = (): number => {
  return Date.now();
};

/**
 * Split a string into host and port
 *
 * @param listen string of format host:port
 */
export const splitListen = (listen: string) =>  {
  const split = listen.split(':');
  return {
    host: split[0],
    port: split[1],
  };
};

/**
 * Get directory of system home.
 */
export const getSystemHomeDir = (): string => {
  switch (os.platform()) {
    case 'win32': return process.env.LOCALAPPDATA!;
    case 'darwin': return path.join(process.env.HOME!, 'Library', 'Application Support');
    default: return process.env.HOME!;
  }
};

/**
 * Get the data directory of a service
 */
export const getServiceDataDir = (service: string) => {
  const homeDir = getSystemHomeDir();
  const serviceDir = service.toLowerCase();

  switch (os.platform()) {
    case 'win32':
    case 'darwin':
      return path.join(homeDir, capitalizeFirstLetter(serviceDir));

    default: return path.join(homeDir, `.${serviceDir}`);
  }
};

export const getOutputType = (type: number) => {
  switch (type) {
    case 0: return OutputType.Bech32;
    case 1: return OutputType.Compatibility;
    default: return OutputType.Legacy;
  }
};

export const getPubKeyHashEncodeFuntion = (outputType: OutputType) => {
  switch (outputType) {
    case OutputType.Bech32:
      return p2wpkhOutput;

    case OutputType.Compatibility:
      return p2shP2wpkhOutput;

    case OutputType.Legacy:
      return p2pkhOutput;
  }
};

export const getScriptHashEncodeFunction = (outputType: OutputType) => {
  switch (outputType) {
    case OutputType.Bech32:
      return p2wshOutput;

    case OutputType.Compatibility:
      return p2shP2wshOutput;

    case OutputType.Legacy:
      return p2shOutput;
  }
};

export const reverseString = (input: string) => {
  return input.split('').reverse().join('');
};

export const reverseBuffer = (input: Buffer) => {
  const buffer = Buffer.allocUnsafe(input.length);

  for (let i = 0, j = input.length - 1; i <= j; i += 1, j -= 1) {
    buffer[i] = input[j];
    buffer[j] = input[i];
  }

  return buffer;
};
