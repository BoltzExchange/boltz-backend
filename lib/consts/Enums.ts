export enum ErrorCodePrefix {
  General = 0,
  Grpc = 1,
  Service = 2,
  Wallet = 3,
  Chain = 4,
  Lnd = 5,
  Swap = 6,
  Rates = 7,
  Backup = 8,
}

export enum ClientStatus {
  Disconnected,
  Connected,
  OutOfSync,
}

export enum OrderSide {
  BUY,
  SELL,
}

export enum SwapUpdateEvent {
  SwapCreated = 'swap.created',
  SwapExpired = 'swap.expired',

  ChannelCreated = 'channel.created',

  InvoiceSet = 'invoice.set',
  InvoicePaid = 'invoice.paid',
  InvoicePending = 'invoice.pending',
  InvoiceSettled = 'invoice.settled',
  InvoiceFailedToPay = 'invoice.failedToPay',

  TransactionFailed = 'transaction.failed',
  TransactionMempool = 'transaction.mempool',
  TransactionClaimed = 'transaction.claimed',
  TransactionRefunded = 'transaction.refunded',
  TransactionConfirmed = 'transaction.confirmed',

  // Events for the prepay miner fee Reverse Swap protocol
  MinerFeePaid = 'minerfee.paid',
}

export enum ChannelCreationType {
  Auto = 'auto',
  Create = 'create',
}

export enum ChannelCreationStatus {
  Attempted = 'attempted',
  Created = 'created',
  Settled = 'settled',
  Abandoned = 'abandoned',
}

export enum ServiceWarning {
  ReverseSwapsDisabled = 'reverse.swaps.disabled',
}

export enum ServiceInfo {
  PrepayMinerFee = 'prepay.minerfee',
}

export enum Network {
  Mainnet = 'mainnet',
  Testnet = 'testnet',
  Simnet = 'simnet',
  Regtest = 'regtest',
}

export enum SwapType {
  Submarine = 'submarine',
  ReverseSubmarine = 'reversesubmarine',
}
