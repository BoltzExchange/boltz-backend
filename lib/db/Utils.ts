import { DatabaseType } from './Database';

type Queries = {
  [DatabaseType.SQLite]: string;
  [DatabaseType.PostgreSQL]: string;
};

export { Queries };
