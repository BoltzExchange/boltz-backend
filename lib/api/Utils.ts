import { Request, Response } from 'express';
import Logger from '../Logger';
import { getHexBuffer } from '../Utils';
import MarkedSwapRepository from '../db/repositories/MarkedSwapRepository';
import CountryCodes from '../service/CountryCodes';
import ServiceErrors from '../service/Errors';
import Errors from './Errors';

type ApiArgument = {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'object';
  hex?: boolean;
  optional?: boolean;
};

// Some endpoints are getting spammed on mainnet, so we don't log the warnings for them
const errorsNotToLog: any[] = [
  ServiceErrors.SWAP_NO_LOCKUP().message,
  ServiceErrors.ETHEREUM_NOT_ENABLED().message,
];

/**
 * Validates that all required arguments were provided in the body correctly
 *
 * @returns the validated arguments
 */
export const validateRequest = (
  body: Record<string, any>,
  argsToCheck: ApiArgument[],
) => {
  const response: any = {};

  argsToCheck.forEach((arg) => {
    const value = body[arg.name];

    if (value !== undefined) {
      if (typeof value === arg.type) {
        if (arg.hex && value !== '') {
          const buffer = getHexBuffer(value);

          if (buffer.length === 0) {
            throw Errors.COULD_NOT_PARSE_HEX(arg.name);
          }

          response[arg.name] = buffer;
        } else {
          response[arg.name] = value;
        }
      } else {
        throw Errors.INVALID_PARAMETER(arg.name);
      }
    } else if (!arg.optional) {
      throw Errors.UNDEFINED_PARAMETER(arg.name);
    }
  });

  return response;
};

export const errorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  error: unknown,
  statusCode = 400,
): void => {
  if (typeof error === 'string') {
    writeErrorResponse(logger, req, res, statusCode, { error });
  } else {
    const errorObject = error as any;

    // Bitcoin Core related errors
    if (errorObject.details) {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.details,
      });
      // Custom error when broadcasting a refund transaction fails because
      // the locktime requirement has not been met yet
    } else if (errorObject.timeoutBlockHeight) {
      writeErrorResponse(logger, req, res, statusCode, error);
      // Everything else
    } else {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.message,
      });
    }
  }
};

export const successResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(200);

  if (typeof data === 'object') {
    res.json(data);
  } else {
    res.write(data);
    res.end();
  }
};

export const createdResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(201).json(data);
};

export const writeErrorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  statusCode: number,
  error: any,
) => {
  if (!errorsNotToLog.includes(error?.error || error)) {
    logger.warn(
      `Request ${req.method} ${req.originalUrl} ${
        req.body && Object.keys(req.body).length > 0
          ? `${JSON.stringify(req.body)} `
          : ''
      }failed: ${JSON.stringify(error)}`,
    );
  }

  setContentTypeJson(res);
  res.status(statusCode).json(error);
};

export const setContentTypeJson = (res: Response) => {
  res.set('Content-Type', 'application/json');
};

export const checkPreimageHashLength = (preimageHash: Buffer) => {
  if (preimageHash.length !== 32) {
    throw `invalid preimage hash length: ${preimageHash.length}`;
  }
};

export const markSwap = async (
  countryCodes: CountryCodes,
  ip: string,
  swapId: string,
) => {
  if (!countryCodes.isRelevantCountry(countryCodes.getCountryCode(ip))) {
    return;
  }

  await MarkedSwapRepository.addMarkedSwap(swapId);
};
