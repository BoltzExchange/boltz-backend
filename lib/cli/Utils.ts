import { Arguments } from 'yargs';
import { address, ECPair, Transaction } from 'bitcoinjs-lib';
import { Networks, detectSwap, constructClaimTransaction, constructRefundTransaction } from 'boltz-core';
import inquire from './inquire';
import { getHexBuffer } from '../Utils';
import { OutputType, OrderSide } from '../proto/boltzrpc_pb';

export const getOrderSide = (side: string) => {
  switch (side.toLowerCase()) {
    case 'buy': return OrderSide.BUY;
    case 'sell': return OrderSide.SELL;

    default: throw `could not find order side: ${side}`;
  }
};

export const getOutputType = (type: string) => {
  switch (type.toLowerCase()) {
    case 'bech32': return OutputType.BECH32;
    case 'compatibility': return OutputType.COMPATIBILITY;

    default: return OutputType.LEGACY;
  }
};

export const getNetwork = (networkKey: string) => {
  const network = Networks[networkKey];

  if (!network) {
    throw `Could not find network: ${networkKey}`;
  }

  return network;
};

const parseParameters = (argv: Arguments) => {
  const network = getNetwork(argv.network);

  return {
    network,
    lockupTransaction: Transaction.fromHex(argv.lockup_transaction),
    redeemScript: getHexBuffer(argv.redeem_script),
    destinationScript: address.toOutputScript(argv.destination_address, network),
  };
};

const parseSwapOutput = (redeemScript: Buffer, lockupTransaction: Transaction) => {
  const swapOutput = detectSwap(redeemScript, lockupTransaction);

  if (!swapOutput) {
    throw 'Could not find swap in transaction';
  }

  return swapOutput;
};

export const parseCommands = async (inquiries: any[], argv: Arguments): Promise<Arguments> => {
  const argvLength = Object.keys(argv).length;
  if (argvLength === inquiries.length) {
    const answers = await inquire(inquiries);
    return { ...answers, ...argv };
  } else {
    return argv;
  }
};

export const claimSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const claimKeys = ECPair.fromPrivateKey(getHexBuffer(argv.claim_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const claimTransaction = constructClaimTransaction(
    {
      redeemScript,
      preimage: getHexBuffer(argv.preimage),
      keys: claimKeys,
    },
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return claimTransaction.toHex();
};

export const refundSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const refundKeys = ECPair.fromPrivateKey(getHexBuffer(argv.refund_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const refundTransaction = constructRefundTransaction(
    refundKeys,
    redeemScript,
    argv.timeout_block_height,
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return refundTransaction.toHex();
};
