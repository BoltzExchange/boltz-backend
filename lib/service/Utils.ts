import { getUnixTime } from '../Utils';
import type { Currency } from '../wallet/WalletManager';
import Errors from './Errors';
import TimeoutDeltaProvider from './TimeoutDeltaProvider';

export const getCurrency = (
  currencies: Map<string, Currency>,
  symbol: string,
) => {
  const currency = currencies.get(symbol);
  if (!currency) {
    throw Errors.CURRENCY_NOT_FOUND(symbol);
  }

  return currency;
};

export const calculateTimeoutDate = (chain: string, blocksMissing: number) => {
  const blockTime = TimeoutDeltaProvider.blockTimes.get(chain);
  if (blockTime === undefined) {
    throw Errors.CURRENCY_NOT_FOUND(chain);
  }

  return getUnixTime() + blocksMissing * blockTime * 60;
};
