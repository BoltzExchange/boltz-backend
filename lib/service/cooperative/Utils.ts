import { crypto } from 'bitcoinjs-lib';
import type { Types } from 'boltz-core';
import {
  createMusig,
  hashForWitnessV1,
  parseTransaction,
  tweakMusig,
} from '../../Core';
import { getHexString } from '../../Utils';
import type Wallet from '../../wallet/Wallet';
import type { Currency } from '../../wallet/WalletManager';
import Errors from '../Errors';
import type { PartialSignature } from './MusigSigner';

export const isPreimageValid = (
  swap: { preimageHash: string },
  preimage: Buffer,
) =>
  preimage.length === 32 &&
  getHexString(crypto.sha256(preimage)) === swap.preimageHash;

export const createPartialSignature = async (
  currency: Currency,
  wallet: Wallet,
  swapTree: Types.SwapTree,
  keyIndex: number,
  theirPublicKey: Buffer,
  theirNonce: Buffer,
  rawTransaction: Buffer | string,
  vin: number,
): Promise<PartialSignature> => {
  const tx = parseTransaction(currency.type, rawTransaction);
  if (vin < 0 || tx.ins.length <= vin) {
    throw Errors.INVALID_VIN();
  }

  const ourKeys = wallet.getKeysByIndex(keyIndex);

  const musig = createMusig(ourKeys, theirPublicKey);
  tweakMusig(currency.type, musig, swapTree);

  musig.aggregateNonces([[theirPublicKey, theirNonce]]);

  const hash = await hashForWitnessV1(currency, tx, vin);
  musig.initializeSession(hash);

  return {
    signature: Buffer.from(musig.signPartial()),
    pubNonce: Buffer.from(musig.getPublicNonce()),
  };
};
