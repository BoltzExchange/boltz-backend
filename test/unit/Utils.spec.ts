// tslint:disable: max-line-length

import os from 'os';
import { OutputType } from 'boltz-core';
import { Transaction } from 'bitcoinjs-lib';
import * as utils from '../../lib/Utils';
import commitHash from '../../lib/Version';
import { OrderSide } from '../../lib/consts/Enums';
import { constructTransaction, randomRange } from '../Utils';

const packageJson = require('../../package.json');

describe('Utils', () => {
  let pairId: string;

  const pair = {
    base: 'BTC',
    quote: 'LTC',
  };

  const sampleTransactions = [
    Transaction.fromHex('01000000017fa897c3556271c34cb28c03c196c2d912093264c9d293cb4980a2635474467d010000000f5355540b6f93598893578893588851ffffffff01501e0000000000001976a914aa2482ce71d219018ef334f6cc551ee88abd920888ac00000000'),
    Transaction.fromHex('010000000001010dabcc426e9f5f57c1000e1560d06ebd21f510c74fe2d0c30fe8eefcabaf31f50200000000fdffffff02a086010000000000160014897ab9fb4e4bf920af9047b5a1896b4689a65bff0e52090000000000160014b3330dbbb43be7a4e2df367f58cf76c74f68141602483045022100ef83bcabb40debd4e13c0eda6b918b940b404344f41253a42c51fc76319ca64502205a1ad86d0bb92f25de75a6b37edf7118863ba9b7c438473ecde07765dedbb9ed012103df1535396d6f4c68458ef3ae86a32e5484e89d7bfc94cabd6c0c09ceaab0b2ab00000000'),
  ];

  test('should split derivation path', () => {
    const master = 'm';
    const sub = [0, 2, 543];
    const derivationPath = `${master}/${sub[0]}/${sub[1]}/${sub[2]}`;

    const result = utils.splitDerivationPath(derivationPath);

    expect(result.master).toEqual(master);
    expect(result.sub.length).toEqual(sub.length);

    sub.forEach((value, index) => {
      expect(result.sub[index]).toEqual(value);
    });
  });

  test('should concat error code', () => {
    const prefix = 0;
    const code = 1;

    expect(utils.concatErrorCode(prefix, code)).toEqual(`${prefix}.${code}`);
  });

  test('should capitalize the first letter', () => {
    const input = 'test123';
    const result = input.charAt(0).toUpperCase() + input.slice(1);

    expect(utils.capitalizeFirstLetter(input)).toEqual(result);
  });

  test('should resolve home', () => {
    const input = '~.boltz';

    if (os.platform() !== 'win32') {
      expect(utils.resolveHome(input).charAt(0)).toEqual('/');
    } else {
      expect(utils.resolveHome(input)).toEqual(input);
    }
  });

  test('should get a hex encoded Buffers and strings', () => {
    const string = 'test';

    expect(utils.getHexBuffer(string)).toEqual(Buffer.from(string, 'hex'));
    expect(utils.getHexString(Buffer.from(string))).toEqual(Buffer.from(string).toString('hex'));
  });

  test('should check whether it is an object', () => {
    expect(utils.isObject({})).toBeTruthy;
    expect(utils.isObject([])).toBeFalsy;
  });

  test('should split host and port', () => {
    const host = 'localhost';
    const port = '9000';

    const input = `${host}:${port}`;

    expect(utils.splitListen(input)).toEqual({ host, port });
  });

  test('should detect whether a transaction signals RBF', () => {
    const inputHash = '9788d1d096dfb41c429a5e76bf2c6e6eb6e3b9aa57feecae3b33c57b4f6fea62';

    expect(utils.transactionSignalsRbfExplicitly(constructTransaction(true, inputHash))).toBeTruthy();
    expect(utils.transactionSignalsRbfExplicitly(constructTransaction(false, inputHash))).toBeFalsy();

    expect(utils.transactionSignalsRbfExplicitly(Transaction.fromHex(sampleTransactions[0].toHex()))).toBeFalsy();
    expect(utils.transactionSignalsRbfExplicitly(Transaction.fromHex(sampleTransactions[1].toHex()))).toBeTruthy();
  });

  test('should generate ids', () => {
    expect(utils.generateId().length).toEqual(6);

    const random = randomRange(10);
    expect(utils.generateId(random).length).toEqual(random);
  });

  test('should get pair ids', () => {
    pairId = utils.getPairId(pair);
    expect(pairId).toEqual('BTC/LTC');
  });

  test('should split pair ids', () => {
    const split = utils.splitPairId(pairId);
    expect(pair.base === split.base && pair.quote === split.quote).toBeTruthy();
  });

  test('should concat error codes', () => {
    const prefix = 0;
    const code = 1;

    expect(utils.concatErrorCode(prefix, code)).toEqual(`${prefix}.${code}`);
  });

  test('should check types of variables', () => {
    expect(utils.isObject([])).toBeFalsy();
    expect(utils.isObject({})).toBeTruthy();
  });

  test('should capitalize the first letter', () => {
    const input = 'boltz';
    const result = input.charAt(0).toUpperCase() + input.slice(1);

    expect(utils.capitalizeFirstLetter(input)).toEqual(result);
  });

  test('should resolve home', () => {
    const input = '~.boltz';

    if (os.platform() !== 'win32') {
      expect(utils.resolveHome(input).charAt(0)).toEqual('/');
    } else {
      expect(utils.resolveHome(input)).toEqual(input);
    }
  });

  test('should convert minutes into milliseconds', () => {
    const random = randomRange(10);
    const milliseconds = random * 60 * 1000;

    expect(utils.minutesToMilliseconds(random)).toEqual(milliseconds);
  });

  test('should get amount of invoice', () => {
    expect(utils.getInvoiceAmt(
      // tslint:disable-next-line: max-line-length
      'lnbcrt100u1pwddnw3pp5rykwp0q399hrcluxnyhv7kfpmk4uttpu00wx9098cesacr9yzk8sdqqcqzpgn9g5vjr0qcudrgu66phz5tx0j0fnxe0gzyl5u6yat9y3xskrqyhherceutcuh9m6h89anphe5un3qac8f2r9j5hykn3uh6z0zkp9racp5lecss',
    )).toEqual(10000);

    expect(utils.getInvoiceAmt(
      // tslint:disable-next-line: max-line-length
      'lnbcrt987650n1pwddnskpp5d4tw4gpjgqdqlgkq5yc309r2kguure53cff8a0kjta5hurltc4yqdqqcqzpgzeu404h9udp5ay39kdvau7m5kdkvycajfhx46slgkfgyhpngnztptulxpx8s7qncp45v5nxjulje5268cu22gxysg9hm3ul8ktrw5zgqcg98hg',
    )).toEqual(98765);

    expect(utils.getInvoiceAmt(
      // tslint:disable-next-line: max-line-length
      'lntltc1u1p0p7fuypp52k9p5mszjc9x0swfw5tum8rsdguzgxmqtg4rr2xaczhkrssvse0sdqqcqzjqxuya98m2834m3xdn7pe6kj5t2q8n3rx9ft37ygvkj07g4a2h7q3xp5mkh44f4jyuupg9t69xf0xknzqa7ufah4dqms5xz4veqplykmcqp4fue6',
    )).toEqual(100);

    expect(utils.getInvoiceAmt(
      // tslint:disable-next-line: max-line-length
      'lntb1p0p7f6lpp58zpsxk88e8uz2lndqrf9radths484jqlhmas92kt7md2cqszpn6qdqqcqzpgrzjqtejjv7p39kcv5gezydzgse8ea3kcw8zqe36afy64zem09us5hjgcxgphsqqqzsqqyqqqqlgqqqqqqgq9q4wqv5a4uwhhvfh93k2ue75lrre50tk99pk689qgf6ul5my5vr749689wcunnv7zjcuk7jlpwz44fv87ra2snsjzw34pnfs5d477u82cpm6cym6',
    )).toEqual(0);
  });

  test('should get rate', () => {
    const rate = 2;
    const reverseRate = 1 / rate;

    expect(utils.getRate(rate, OrderSide.BUY, true)).toEqual(reverseRate);
    expect(utils.getRate(rate, OrderSide.SELL, true)).toEqual(rate);

    expect(utils.getRate(rate, OrderSide.BUY, false)).toEqual(rate);
    expect(utils.getRate(rate, OrderSide.SELL, false)).toEqual(reverseRate);
  });

  test('should the chain currency', () => {
    const { base, quote } = pair;

    expect(utils.getChainCurrency(base, quote, OrderSide.BUY, true)).toEqual(base);
    expect(utils.getChainCurrency(base, quote, OrderSide.SELL, true)).toEqual(quote);

    expect(utils.getChainCurrency(base, quote, OrderSide.BUY, false)).toEqual(quote);
    expect(utils.getChainCurrency(base, quote, OrderSide.SELL, false)).toEqual(base);
  });

  test('should the lightning currency', () => {
    const { base, quote } = pair;

    expect(utils.getLightningCurrency(base, quote, OrderSide.BUY, true)).toEqual(quote);
    expect(utils.getLightningCurrency(base, quote, OrderSide.SELL, true)).toEqual(base);

    expect(utils.getLightningCurrency(base, quote, OrderSide.BUY, false)).toEqual(base);
    expect(utils.getLightningCurrency(base, quote, OrderSide.SELL, false)).toEqual(quote);
  });

  test('should get output type', () => {
    expect(utils.getOutputType(0)).toEqual(OutputType.Bech32);
    expect(utils.getOutputType(1)).toEqual(OutputType.Compatibility);
    expect(utils.getOutputType(2)).toEqual(OutputType.Legacy);

    expect(utils.getOutputType()).toEqual(OutputType.Legacy);

    expect(() => utils.getOutputType(123)).toThrow(new Error('type does not exist'));
  });

  test('should get memo for swaps', () => {
    expect(utils.getSwapMemo('LTC', false)).toBe('Send to LTC lightning');
    expect(utils.getSwapMemo('BTC', true)).toBe('Send to BTC address');
  });

  test('should get sending and receiving currency', () => {
    const baseCurrency = 'LTC';
    const quoteCurrency = 'BTC';

    expect(utils.getSendingReceivingCurrency(baseCurrency, quoteCurrency, OrderSide.BUY)).toEqual({
      sending: baseCurrency,
      receiving: quoteCurrency,
    });

    expect(utils.getSendingReceivingCurrency(baseCurrency, quoteCurrency, OrderSide.SELL)).toEqual({
      sending: quoteCurrency,
      receiving: baseCurrency,
    });
  });

  test('should format errors', () => {
    const test = 'error';
    const object = { test };
    const objectMessage = { message: test };

    expect(utils.formatError(test)).toEqual(test);
    expect(utils.formatError(object)).toEqual(JSON.stringify(object));
    expect(utils.formatError(objectMessage)).toEqual(test);
  });

  test('should get version', () => {
    expect(utils.getVersion()).toEqual(`${packageJson.version}${commitHash}`);
  });
});
