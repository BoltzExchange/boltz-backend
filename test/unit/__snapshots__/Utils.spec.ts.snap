// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils stringify should stringify bigint 1`] = `
"{
  "big": "1"
}"
`;

exports[`Utils stringify should stringify objects 1`] = `
"{
  "test": "data"
}"
`;
